### ISSUE TITLE
ERROR ~ Error executing process > 'PIPELINE:SEROTYPE (NP-0087-IDRL-AKU_S92_trimmed)'

### ISSUE THREAD

**Issue by Samiah-Kanwar** [HarryHung/gps-unified-pipeline#52]
Hello Harry, while executing the `gps-unified-pipeline` on a batch of 12 samples (FASTQ sequences), I encountered following error on one of my sample: 

```
ERROR ~ Error executing process > 'PIPELINE:SEROTYPE (NP-0087-IDRL-AKU_S92_trimmed)'

Caused by:
  Process `PIPELINE:SEROTYPE (NP-0087-IDRL-AKU_S92_trimmed)` terminated with an error exit status (1)

Command executed:

  SEROBA_DIR="seroba"
  DATABASE="database"
  READ1="processed-NP-0087-IDRL-AKU_S92_trimmed_1.fastq.gz"
  READ2="processed-NP-0087-IDRL-AKU_S92_trimmed_2.fastq.gz"
  SAMPLE_ID="NP-0087-IDRL-AKU_S92_trimmed"
  
  source get_serotype.sh

Command exit status:
  1

Command output:
  cluster detected 1 threads available to it
  cluster reported completion
  cluster_3 detected 1 threads available to it
  cluster_3 reported completion
  cluster_4 detected 1 threads available to it
  cluster_4 reported completion
  cluster_6 detected 1 threads available to it
  cluster_6 reported completion
  
  0.15055493003056136
  /seroba-1.0.2/build/kmc_tools simple temp.kmco93qiglx/NP-0087-IDRL-AKU_S92_trimmed seroba/database/kmer_db/Swiss_NT/Swiss_NT intersect temp.kmco93qiglx/inter
  0.0
  /seroba-1.0.2/build/kmc_tools simple temp.kmco93qiglx/NP-0087-IDRL-AKU_S92_trimmed seroba/database/kmer_db/43/43 intersect temp.kmco93qiglx/inter
  0.01530841963079694
  /seroba-1.0.2/build/kmc_tools simple temp.kmco93qiglx/NP-0087-IDRL-AKU_S92_trimmed seroba/database/kmer_db/39X/39X intersect temp.kmco93qiglx/inter
  0.03521542396222703
  /seroba-1.0.2/build/kmc_tools simple temp.kmco93qiglx/NP-0087-IDRL-AKU_S92_trimmed seroba/database/kmer_db/25A/25A intersect temp.kmco93qiglx/inter
  0.023278764658075005
  /seroba-1.0.2/build/kmc_tools simple temp.kmco93qiglx/NP-0087-IDRL-AKU_S92_trimmed seroba/database/kmer_db/17A/17A intersect temp.kmco93qiglx/inter
  0.02832162954006418
  /seroba-1.0.2/build/kmc_tools simple temp.kmco93qiglx/NP-0087-IDRL-AKU_S92_trimmed seroba/database/kmer_db/47F/47F intersect temp.kmco93qiglx/inter
  0.0592279596075342
  /seroba-1.0.2/build/kmc_tools simple temp.kmco93qiglx/NP-0087-IDRL-AKU_S92_trimmed seroba/database/kmer_db/10A/10A intersect temp.kmco93qiglx/inter
  0.018366189193022266
  /seroba-1.0.2/build/kmc_tools simple temp.kmco93qiglx/NP-0087-IDRL-AKU_S92_trimmed seroba/database/kmer_db/28F/28F intersect temp.kmco93qiglx/inter
  0.12212593297600562
  /seroba-1.0.2/build/kmc_tools simple temp.kmco93qiglx/NP-0087-IDRL-AKU_S92_trimmed seroba/database/kmer_db/37/37 intersect temp.kmco93qiglx/inter
  0.0
  /seroba-1.0.2/build/kmc_tools simple temp.kmco93qiglx/NP-0087-IDRL-AKU_S92_trimmed seroba/database/kmer_db/11C/11C intersect temp.kmco93qiglx/inter
  0.0007471607890017932
  /seroba-1.0.2/build/kmc_tools simple temp.kmco93qiglx/NP-0087-IDRL-AKU_S92_trimmed seroba/database/kmer_db/42/42 intersect temp.kmco93qiglx/inter
  0.035696073431922486
  15C
  {'15A': 0, '15B': 0, '15C': 0, '15F': 16}
  15A
  {'genes': [], 'pseudo': [], 'allele': [], 'snps': []}
  15B
  {'genes': [], 'pseudo': [], 'allele': [], 'snps': []}
  15C
  {'genes': [], 'pseudo': [], 'allele': [], 'snps': []}
  15C
  15F
  {'genes': [], 'pseudo': [], 'allele': [], 'snps': []}
  {'15A': -1, '15B': 0, '15C': -2.5, '15F': 15}
  {'15A': {'genes': [], 'pseudo': [], 'allele': [], 'snps': []}, '15B': {'genes': [], 'pseudo': [], 'allele': [], 'snps': []}, '15C': {'genes': [], 'pseudo': ['wciZ'], 'allele': [], 'snps': []}, '15F': {'genes': [], 'pseudo': [], 'allele': [], 'snps': []}}
  ['15A', '15C']
  15A
  15B/15C
  15C
  None

Command error:
  0.01530841963079694
  /seroba-1.0.2/build/kmc_tools simple temp.kmco93qiglx/NP-0087-IDRL-AKU_S92_trimmed seroba/database/kmer_db/39X/39X intersect temp.kmco93qiglx/inter
  0.03521542396222703
  /seroba-1.0.2/build/kmc_tools simple temp.kmco93qiglx/NP-0087-IDRL-AKU_S92_trimmed seroba/database/kmer_db/25A/25A intersect temp.kmco93qiglx/inter
  0.023278764658075005
  /seroba-1.0.2/build/kmc_tools simple temp.kmco93qiglx/NP-0087-IDRL-AKU_S92_trimmed seroba/database/kmer_db/17A/17A intersect temp.kmco93qiglx/inter
  0.02832162954006418
  /seroba-1.0.2/build/kmc_tools simple temp.kmco93qiglx/NP-0087-IDRL-AKU_S92_trimmed seroba/database/kmer_db/47F/47F intersect temp.kmco93qiglx/inter
  0.0592279596075342
  /seroba-1.0.2/build/kmc_tools simple temp.kmco93qiglx/NP-0087-IDRL-AKU_S92_trimmed seroba/database/kmer_db/10A/10A intersect temp.kmco93qiglx/inter
  0.018366189193022266
  /seroba-1.0.2/build/kmc_tools simple temp.kmco93qiglx/NP-0087-IDRL-AKU_S92_trimmed seroba/database/kmer_db/28F/28F intersect temp.kmco93qiglx/inter
  0.12212593297600562
  /seroba-1.0.2/build/kmc_tools simple temp.kmco93qiglx/NP-0087-IDRL-AKU_S92_trimmed seroba/database/kmer_db/37/37 intersect temp.kmco93qiglx/inter
  0.0
  /seroba-1.0.2/build/kmc_tools simple temp.kmco93qiglx/NP-0087-IDRL-AKU_S92_trimmed seroba/database/kmer_db/11C/11C intersect temp.kmco93qiglx/inter
  0.0007471607890017932
  /seroba-1.0.2/build/kmc_tools simple temp.kmco93qiglx/NP-0087-IDRL-AKU_S92_trimmed seroba/database/kmer_db/42/42 intersect temp.kmco93qiglx/inter
  0.035696073431922486
  15C
  {'15A': 0, '15B': 0, '15C': 0, '15F': 16}
  15A
  {'genes': [], 'pseudo': [], 'allele': [], 'snps': []}
  15B
  {'genes': [], 'pseudo': [], 'allele': [], 'snps': []}
  15C
  {'genes': [], 'pseudo': [], 'allele': [], 'snps': []}
  15C
  15F
  {'genes': [], 'pseudo': [], 'allele': [], 'snps': []}
  {'15A': -1, '15B': 0, '15C': -2.5, '15F': 15}
  {'15A': {'genes': [], 'pseudo': [], 'allele': [], 'snps': []}, '15B': {'genes': [], 'pseudo': [], 'allele': [], 'snps': []}, '15C': {'genes': [], 'pseudo': ['wciZ'], 'allele': [], 'snps': []}, '15F': {'genes': [], 'pseudo': [], 'allele': [], 'snps': []}}
  ['15A', '15C']
  15A
  15B/15C
  15C
  None
  Traceback (most recent call last):
    File "/usr/local/bin/ser
![image](https://github.com/HarryHung/gps-unified-pipeline/assets/87472403/973c6172-850f-47d4-badb-061373d7c6df)
oba", line 4, in <module>
      __import__('pkg_resources').run_script('seroba==1.0.2', 'seroba')
    File "/usr/lib/python3/dist-packages/pkg_resources/__init__.py", line 658, in run_script
      self.require(requires)[0].run_script(script_name, ns)
    File "/usr/lib/python3/dist-packages/pkg_resources/__init__.py", line 1445, in run_script
      exec(script_code, namespace, namespace)
    File "/usr/local/lib/python3.6/dist-packages/seroba-1.0.2-py3.6.egg/EGG-INFO/scripts/seroba", line 86, in <module>
    File "/usr/local/lib/python3.6/dist-packages/seroba-1.0.2-py3.6.egg/seroba/tasks/sero_run.py", line 19, in run
    File "/usr/local/lib/python3.6/dist-packages/seroba-1.0.2-py3.6.egg/seroba/serotyping.py", line 481, in run
    File "/usr/local/lib/python3.6/dist-packages/seroba-1.0.2-py3.6.egg/seroba/serotyping.py", line 453, in _prediction
    File "/usr/local/lib/python3.6/dist-packages/seroba-1.0.2-py3.6.egg/seroba/serotyping.py", line 397, in _find_serotype
  TypeError: argument of type 'NoneType' is not iterable

Work dir:
  /home/samiah/Desktop/Desktop/AKU_System/gps-unified-pipeline/work/18/310a53cf47065e514606a58c8a7669

Tip: you can replicate the issue by changing to the process work dir and entering the command `bash .command.run`

 -- Check '.nextflow.log' file for details

```
The error shows that the SeroBA failed to assigned the serotype to this particular sample but I am unable to understand the reason. Moreover the pipeline halted without creating the final `result.tsv` file. Is there any way to get the AMR analysis, MLST and other analysis that executed successfully?

**Comment by HarryHung** [comment]
> Is there any way to get the AMR analysis, MLST and other analysis that executed successfully?


It is not possible at the moment, but not impossible for me to implement this function.

-----

If you remove this sample from the input, can the pipeline be successfully completed?

Is it possible for you to share that sample with me? You can send it to ch31@sanger.ac.uk.

**Comment by Samiah-Kanwar** [comment]
> If you remove this sample from the input, can the pipeline be successfully completed?

yes, when I removed the sample, the pipeline ran successfully.

**Comment by HarryHung** [comment]
Thanks for the confirmation. 
If you can share the sample with me, I can check whether it is fixable.
If it is not fixable from the pipeline prespective (e.g. a bug of SeroBA), I can update the pipeline to handle the error gracefully.

**Comment by Samiah-Kanwar** [comment]
> Thanks for the confirmation. If you can share the sample with me, I can check whether it is fixable. If it is not fixable from the pipeline prespective (e.g. a bug of SeroBA), I can update the pipeline to handle the error gracefully.

Sure, thank you very much. I will share the sample with you via email shortly.

**Comment by HarryHung** [comment]
Thanks for the samples, in my test, it seems to be a bug of SeroBA:
https://github.com/sanger-pathogens/seroba/issues/73

I have reported it and working with the maintainers of SeroBA to determine whether a fix can be implemented and pushed out.

**Comment by HarryHung** [comment]
@Samiah-Kanwar since I am unsure whether SeroBA issue will be fixed despite knowing the root cause, I have implemented SeroBA failure protection at the ["feature/seroba-failure-protection" branch](https://github.com/HarryHung/gps-unified-pipeline/tree/feature/seroba-failure-protection)

If this branch works well for you, I can push it as a new update ASAP.

**Comment by Samiah-Kanwar** [comment]
Thank you very much Harry for your support. I have seen your reply on
github issue I will re-run the sample and get back to you.

kind regards,
Samiah Kanwar

**Comment by Samiah-Kanwar** [comment]
Hi Harry, I tried but facing the same error.

**Comment by HarryHung** [comment]
Can you confirm you are testing with the `feature/seroba-failure-protection` branch (i.e. check with `git branch -va`)? I have not merged that branch yet. 

Also SeroBA has a pending release that fixes this particular error. I will integrate that ASAP as well.

**Comment by Samiah-Kanwar** [comment]
> Can you confirm you are testing with the `feature/seroba-failure-protection` branch (i.e. check with `git branch -va`)? I have not merged that branch yet.

Yes, I am testing with `feature/seroba-failure-protection` branch 

![image](https://github.com/HarryHung/gps-unified-pipeline/assets/87472403/39043c86-11d9-4532-87c9-f2f51e3eede9)

>Also SeroBA has a pending release that fixes this particular error. I will integrate that ASAP as well.

That would be great. Thanks!

**Comment by HarryHung** [comment]
The printout of `git branch -va` indicates you are still working with the `master` branch (the active branch is with an asterisk * symbol)
Please switch to the `feature/seroba-failure-protection` branch by running `git switch feature/seroba-failure-protection`

**Comment by Samiah-Kanwar** [comment]
> Please switch to the `feature/seroba-failure-protection` branch by running `git switch feature/seroba-failure-protection`

Thank you for pointing this out, I have switched the branch and got the output.

![image](https://github.com/HarryHung/gps-unified-pipeline/assets/87472403/082c1fad-8316-49f6-825a-3be1617afec2)

You can push it as new update. Once again many thanks for your support.

**Comment by HarryHung** [comment]
Thanks for the confirmation. I will see whether it is possible to wait for the next release of SeroBA as well.

**Comment by HarryHung** [comment]
Issue resolved in [release v0.8.2](https://github.com/HarryHung/gps-unified-pipeline/releases/tag/v0.8.2-beta)
